### PR TITLE
[jasmine] Fix buggy typings with overload resolution

### LINKED ISSUE #42455: @types/jasmine toHaveBeenCalledWith infers parameters for overloads incorrectly
[The function signature for `toHaveBeenCalledWith`](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/7456aa4eec4bcfc03259a7d32e619d0607bbed3b#r37354130) unfortunately breaks if a generic function that consists of overloads is passed. TypeScript is not able to infer the proper overload for functions through generics. Only a single overload signature is considered for parameter inferring. For example:

```ts
function hello(idx: number): void;
function hello(item: object): void;
function hello(arg: any): void { /* impl */}

// This won't work because `MatchableArgs` only infers `object` as parameter type.
expect(hello).toHaveBeenCalledWith(1);
```

This is a [known limitation in TypeScript](https://github.com/microsoft/TypeScript/issues/32418), but what I'd love to see here is a way to work around this limitation. Would it be helpful to provide a generic to `toHaveBeenCalledWith` that can be used to specify parameters manually if needed?

Authors: @zvirja @djungowski @kolodny

### PR DESCRIPTION
**Pull request by chivesrs**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: See below.
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

Fixes #42455 and #39162. Apparently parameter inference with overloads does not work very well per https://github.com/microsoft/TypeScript/issues/32418.

I've added typings tests in both the outer and inner `ts3.1` directories to make sure. Here's a Codepen to prove everything works at runtime (it's slightly different than the typings tests but it proves the same point): https://codepen.io/chivesrs/pen/dyoJywz.

### CODE REVIEW

**Comment by chivesrs** [comment]
Fyi, I didn't request a review from Boris, that's GitHub doing GitHub things.
Feel free to merge whenever.

**Review by devoto13 — CHANGES_REQUESTED** [review]
I'm in favour of reverting to `any` as it is better to have looser types over having types blocking valid code. But I think this deserves some discussion. In particular I would like to hear what @zvirja (who introduced these changes) thinks before merging this. Can we somehow provide a fallback to `any` if inference fails?

Sorry, but I'll block this for now.

**Comment by chivesrs** [comment]
@devoto13 I don't mind waiting for the original author.
But, let me give you my thoughts.

If you are a user of these typings, and you find that they fail to resolve on overloaded types, your first intuition is you did something wrong, because given such a popular package/typings these are, they obviously can't be wrong.

The current workaround is to manually type every single thing as I believe `NonTypedSpyObj` (or something similarly named). Our code base at work (Google) has thousands and thousands of lines of code that break due to this assumption. I'm sure many other projects have the same issue.

Also, @zvirja was mentioned in the underlying issue, which is 20 days old, and never responded. Again, for a popular package like this, we shouldn't have stale issues without the maintainers adding some sort of comment.

**Comment by chivesrs** [comment]
Also, in #38412, you see many examples of people being broken by these types, it seems one of them decided to just pin their project on an older release of the typings. I'm not sure that's a behavior we want to promote.

**Comment by chivesrs** [comment]
And my other thoughts on this is, you as a user shouldn't be required to dig into the GitHub issues to figure out that the current workaround is `NonTypedSpyObj` (or whatever it is). Everything that works at runtime without being obviously wrong in some way should also work at compile time with TypeScript.

**Comment by chivesrs** [comment]
I also found this doc in the official TypeScript JS -> TS migration docs
https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html#too-manytoo-few-arguments

that specifically shows people how to do overloading. As a basic JS -> TS migration user, this particular issue in Jasmine would block them, as they're just learning how TypeScript works. This  doesn't lead to a good new user experience.

**Comment by devoto13** [comment]
> If you are a user of these typings, and you find that they fail to resolve on overloaded types, your first intuition is you did something wrong, because given such a popular package/typings these are, they obviously can't be wrong.

I actually have this feeling regularly with libraries much more popular than `@types/jasmine`. So while I share your desire to live in the perfect world, unfortunately problems happen and we need to investigate, report and fix them. The only way to not break anything is to not do anything.

> I also found this doc in the official TypeScript JS -> TS migration docs
https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html#too-manytoo-few-arguments

My personal stance on this is that this is a bad practice to have such functions, but I realise that others may not agree and that there is a lot of such code in a wild.

---

I'm convinced by your other argument, that we should not require a special workaround for valid code to be type-checked. Would it make sense to invert the logic then? 

The default `SpyObj` would not perform any extra type checking and therefore will work out of the box, providing weaker types. Make `NonTypedSpyObj` an alias of `SpyObj` to not break existing usage. Then introduce a `TypedSpyObj`, which does enable type checking for parameters and return types. This way we don't throw away existing work and users can opt in (as I think strict types are useful for many cases). Then `TypedSpyObj` can evolve over time and once robust enough become a default.

**Comment by kolodny** [comment]
This is a limitation of TS, the spy function basically uses the `Parameters` and `ReturnType` types which don't really work as expected for overloads if the last overload isn't correctly encompassing the previous ones

http://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABAWwJ4DFwQBQA8BciYIyARgKYBOAlIcWVQNwBQoksCKGWehAzlEowwAcwA0iVP0HCRtRAKGiWbaPCRpMkbADo9AB0IBDMKgDaAXXn0KlRAB8FM0YgDezRJ8SVyUEJSQAFhYAX2ZmABtfRH0jSiNkPkIABTiE3yo+AB4oVH1yOGAuLQgAPkZEAHpKxAQI1EQoIwBrcj5GgAtyRAijAVqANyoIuCMAE0jonyhCACVffzAAFTzyHNXC4qxSqpq6hqbW9qgunr6oQeHRiaA

[This has been brought up before](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34080) and comes down to a design limitation of the language for overloaded functions 
- https://github.com/microsoft/TypeScript/issues/30545
- https://github.com/microsoft/TypeScript/pull/21496#issuecomment-367959351

It seems from the linked issue that the advised suggestion from the TS team is to ensure the last overloaded term covers the other overloads

> When inferring from a type with multiple call signatures (such as the type of an overloaded function), inferences are made from the last signature (which, presumably, is the most permissive catch-all case). It is not possible to perform overload resolution based on a list of argument types (this would require us to support typeof for arbitrary expressions, as suggested in #6606, or something similar).

Ultimately it comes down to a tradeoff, the same one that the TS team used for the built in `Parameters` and `ReturnType` types which when used on overloaded terms use the last term and don't fallback to considering them `any` which this PR is doing.

//cc @RyanCavanaugh in case anything has changed in TS to make this possible

With all that in mind I'd vote to keeping the current typings as is and pointing folks to the TS issues that cover this (and potentially this comment as well)

**Comment by chivesrs** [comment]
@devoto13 I think inverting them may be a good compromise, that way users who want this behavior can get it. I'll look into calling it "StrictTypedSpyObj" or something. Give me a bit to make the changes.

@kolodny I see no point in using overloads if the last one is the most permissive that is a combination of all of them. For example, if you look at the example in my original code, if I added the overload of `myFunc(x: string|number, y: number)`, this would be the most permissive combo, but it would introduce a new combo of parameters that are not intended. Given that, I don't think we should assume that a last most permissive combo exists (and all the implications that has on this issue).

**Comment by devoto13** [comment]
@chivesrs Thanks! But maybe don't rush with this. It may be a good idea to get :+1: from other maintainers that this is the solution we want to have.

**Comment by chivesrs** [comment]
Sounds good.
I'll give it a day or two or five until I roll with the `StrictTypedSpyObj` approach, and swap the default out to the non strict version.

**Comment by chivesrs** [comment]
@devoto13 I'm going to go with the "swap the default strict and opt-in not strict" approach to be default not strict and opt-in strict. That way basic users of the language won't need to do mass amounts of debugging to figure out why overload spying isn't working.

Also referring non-goal #3 in https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals regarding a balance between correctness and productivity.

**Comment by devoto13** [comment]
@chivesrs Sorry for the late reply, got busy at work. 

Re-reading this and linked threads it seems that most of the errors people reported with strict types are actually valid errors and either have been fixed in Jasmine types or fixed by reporters.

The overloads is a tricky issue as this is a TypeScript limitation and by "fixing" it we'll disable type safety for a lot of other cases. The ratio between what we loose and what we gain is why I wanted to have a discussion and opinions from other maintainers. Second reason is the consensus, so somebody won't come in a couple of months and won't re-revert your changes same as you wanted to revert @zvirja's without much discussion.

In my projects stricter types found several non-critical errors, so minor gain. I also prefer to write very trivial code, so didn't face any pain when updating to the stricter types. I'm on the fence on what is the right thing to do here: leave stricter types by default or change to opt-in (from TS issue it does not seem like overloads will be fixed in the foreseeable future, so doing later is comparable to removing stricter types).

Having said above I have only slight preference towards keeping stricter types. You and @kolodny seems to have opposite opinions. As my opinion is not strong, I'll retract my review now and let others decide whether they care enough to approve or reject this change.

**Comment by kolodny** [comment]
@devoto13 I think I was unclear. I'm very much in agreement with what you said and believe we should keep the strict typings as is. I strongly agree that making spy return values untyped would be a net negative and as such, don't believe we should make that change.

I think that the way it currently is catches a lot of typing errors with spies and mocks and in the case of overloads can be easily overcome with a simple cast, if we were to remove these typings there wouldn't be a __simple__ way to opt into them.

If it comes down to a tallying vote, consider me in the keep things as is camp

**Comment by chivesrs** [comment]
All right.
I think I can agree with the net gain discussion.

**Comment by chivesrs** [comment]
I'll see if I can add some comments in the typings source code somewhere regarding overloads. New PR for that.

**Pull request closed without merging**
